### PR TITLE
Background traffic distribution

### PR DESCRIPTION
**Pull request by mskpyn**
- Added protocol distribution percentage for background traffic to Petabi rule
- Modified Sniffles to create background traffic with given protocol distribution from Petabi rule

### CODE REVIEW

File: sniffles/petabi_rule_writer/petabi_rule_writer.py [hunk 1]
```diff
@@ -185,16 +196,66 @@ def check_pcre_compile(re):
...
+    i = 0
+    while i < len(protocol_list):
+        if protocol_dist[i] != "*":
+            protocol_percent[protocol_list[i]] = int(protocol_dist[i])
+        i += 1
+
+    # Change ratio to percentage for protocol if they don't add up to 100
+    if len(protocol_list) == len(protocol_percent):
+        protocol_sum = sumPercentage(protocol_percent)
+        while protocol_sum != 100:
```
Comment (vvalgenti) [inline sniffles/petabi_rule_writer/petabi_rule_writer.py]: 1. I think we can calculate the protocol sum above.
2. If the sum is too small, just add the remainder to the first entry in the list, or redistribute the remaining evenly amongst all others, or, perhaps better, Error and warn the user.  Silently adjusting might cause errors the user does not expect.
3. Same if it is too big.

File: sniffles/sniffles.py [hunk 2]
```diff
@@ -177,22 +186,36 @@ def start_generation(sconf):
...
+                if back_traffic_distribution:
+                    protocol_pick = random.randint(0, 99)
+                    protocol = None
+                    pick_sum = 0
+                    for proto in back_traffic_distribution:
+                        pick_sum += back_traffic_distribution[proto]
+                        if (protocol_pick < pick_sum):
+                            protocol = proto
+                            break
+                    if protocol is None:
```
Comment (vvalgenti) [inline sniffles/sniffles.py]: If we have a distribution, why would we ever arrive at a point where we don't get a protocol?

File: sniffles/petabi_rule_writer/petabi_rule_writer.py [hunk 3]
```diff
@@ -185,16 +196,66 @@ def check_pcre_compile(re):
...
+# Label protocol percentage with given distribution.
+# Wildcard (* symbol) is used to denote even percentage among wildcards.
+# Wildcard protocol will not be shown in the rule, but will be applied
+# in sniffles. When all protocols are given a percentage and their sum
+# does not add up to 100, they are treated as ratio.
+# Output: Dictionary containing protocol percentage
+def protocolPercentage(protocol_dist):
+    protocol_percent = OrderedDict()
+    total = 0
+    num_wildcard = 0
```
Comment (vvalgenti) [inline sniffles/petabi_rule_writer/petabi_rule_writer.py]: I'm not seeing where wildcards are handled, specifically.  I think the extra code below (where my other comment is) is accomplishing that.  But probably need to make that more clear.  Also, if this variable is not needed, I think we can remove it.

File: sniffles/rulereader.py [hunk 4]
```diff
@@ -1491,6 +1548,7 @@ def readRuleFile(self, filename):
             else:
                 self.all_rules = parser.getRules()
 
+
```
Comment (mskpyn) [inline sniffles/rulereader.py]: Accidentally added an empty line, will remove it in next commit.

File: sniffles/petabi_rule_writer/petabi_rule_writer.py [hunk 5]
```diff
@@ -69,6 +71,17 @@ def main():
         elif opt == "-c":
             if arg is not None:
                 count = arg
+        elif opt == "-D":
+            if arg is not None:
+                protocol_dist = re.split('[\s,;]*', arg)
```
Comment (vvalgenti) [inline sniffles/petabi_rule_writer/petabi_rule_writer.py]: Do we want a star in this regex?  Wouldn't a + be better?

**Pull request merged**